### ISSUE TITLE
Memoized?

### ISSUE THREAD

**Issue by 23doors** [thadeusb/flask-cache#16]
I am a little worried with how memoized seems to work here. Say, we're using more persistent cache than simple thread local storage.

Specifically if we use the example here:
        @cache.memoize(3600)
        def has_membership(role):
                return self.groups.filter_by(role=role).count() >= 1

And call has_membership('something') - it is cached.

In following request we call: cache.delete_memoized('has_membership') and what it should do is delete all memoized versions of this function. Well, it won't. 

Correct me if I am wrong but you're keeping track of memoized functions by simply storing them in self field and in following requests - it is empty on init. This is not scalable at all and will only work in simplest of cases. Bad design decision, that's for sure.

**Comment by thadeusb** [comment]
~~Wrong. It will work as intended.~~

On the call to cache.delete_memoized the cache_key is computed again and used to delete the cache from the backing store.--

self._memoized is only used as a convenience for `get_memoized_names` and `get_memoized_keys`, and to be honest I do not believe those two functions are actually being used anywhere, so perhaps they should be deprecated since they would not work as intended in production?

**Comment by 23doors** [comment]
Really?
It seems like:
1) delete_memoized is running:
self._memoized[:] = [x for x in self._memoized if not deletes(x)]
which doesnt do anything if self._memoized is empty.

2) cache_key is based on: function_name, args and kwargs. So no way to deduce cache_key and delete all memoized functions by simply knowing function_name.

**Comment by thadeusb** [comment]
I see it now. Did you have anything in mind for a solution?

**Comment by thadeusb** [comment]
It used to be able to delete all memoized if just the function name was provided, however since the md5/base64 modifications to the cache_key this capability was lost.

Would it make sense to store what is in self._memoized inside the cache instead? I think this would provide the cross-threading support required. That or the cache_key for memoization must contain the function name, and only the arguments/kwargs are md5'd

**Comment by 23doors** [comment]
You cant save self._memoized either as without locking, there is no way to ensure that everything gets saved there.

E.g. thread #1 wants to add new memoized funnction. Thread #2 got the same idea.
#1 gets current _memoized from memcached, #2 as well. #1 adds his function, #2 adds his. They both save. Only one version will persist and it will contain only memoized function from #1 or #2.

As we're dealing with very fast mechanisms here anyway, there is pretty much only one good way to do it. Something like django johnny-cache does:

when memoizing:

```
version_data = cache.get(function_name)
if version_data is None:
    generate_new_random_md5
    cache.set(function_name, new_version_data)
data = cache.get(function_name + args_kwargs_hash + version_data)
if data is None:
    data = run_function and save new cache
return data 
```

Now delete memoized simply does:
    if args/kwargs - get version_data, delete function_name + args_kwargs_hash + version_data
    else: set version_data to new_random_md5_hash

Basically data keys become something like:
function_name_d2256a05b455fc17bc13e22055ac969b.609c6f0c04906544153056336eff267d

You have to take into account memcache max key here as well. But no need for 32 char length version data. You can just use 6-8 chars.

**Comment by thadeusb** [comment]
If you agree that the latest commit solves this issue, I will release version 0.5.

**Comment by 23doors** [comment]
Looks good, haven't tested it, but everything looks fine and I'm sure you tested it :)

One suggestion though:
from line 347:

```
        if not args and not kwargs:
            version_key = self._memvname(fname)
            self.cache.delete(version_key)
```

Here it may be a good idea to use set(new_version_key) instead of delete. That's how e.g. johnny-cache uses it. Reasons behind that are simple: cache.delete is slow for memcached, cache.set is much faster. Don't know about other backends, but possibly it is true as well.

Also, it may be good idea to not allow timeout=0 for memoized. Or at least point it out in documentation somewhere. Timeout=0 here would mean that old memoized data (that are no longer reachable due to changed version_key) will stay in cache forever.

This may not be true for some library (even with memcached) as e.g. cache libraries werkzeug uses for memcache are chosen from first available of: pylibmc, GAE memcached, python-memcached. Don't know about GAE version but from what I remember, pylibmc treats 0 as 0 seconds instead of forever as memcached server and python-memcached library does.

Maybe simply set timeout to something like 1 month, if 0 is passed (just for memoize).

While we're at memoize - it may be useful to provide new argument for it - 'name' or something like that. It would default to `f.__name__` and can be used as string or as a callable. If callable, it will expect it to accept one parameter and will pass `f.__name__`.

It is easy to add: make memoize accept it, default to None, check within decorator if it is None then use f.**name**. Possibly save it to decorated_function as make_name. Within memoize inner function check if it is callable then use make_name(f.**name**), otherwise simply use make_name - but this logic possibly shouldn't be in memoize inner function but in memoize_make_cache_key.

Why is it needed?
So we can have something like this:

```
class Object:
    switch = False

    """some memoized function here that depends on self.switch - it changes its output completely"""
```

Now if we use name=lambda x: '%s_%s' % (x, self.switch) - we got it working without having to create separately named function for it. Or maybe I am missing something and it is already possible here? (but it seems to be possible for @cached only).

Some way along the line, I would hope to have flask-cache as generic as can be. To eventually support for example this very specific scenario. Imagine this situation - we have a database with millions of Message objects. Messages are grouped into Events. Each message has event_id field.

Now we want to cache per-event messages and we want to invalidate them ONLY if someone manipulates messages (insert/update/delete) with this event_id (not global table invalidation).
We could achieve this by:
memoize get and all other SELECTing functions with key based on: func.**name** + event_id_version_key + global_table_version_key (not function related). INSERT/DELETE/UPDATE would do: cache.set(event_id_version_key, new_random_key), while changing table DDL (or operation that is not event_id specific) would do: cache.set(global_table_version_key, new_random_key).

This is much, much smarter than johnny-cache (which is as easy as it gets - simply invalidates whole table every time) and Flask is currently missing out on some more advanced caching scenarios.

One last thing:
line: 217, it would be great if it wouldn't treat same calls of func(first,second) differently e.g. func(1,0) and func(1,second=0) and func(first=1,second=0).
One way to achieve that - convert all to kwargs or all to args (easiest).

I've done something similar some time ago. Basically it comes down to:

```
new_args = []
arg_num = 0
m_args = inspect.getargspec(func)[0]

for i in range(len(m_args)):
    if m_args[i] in kwargs:
        new_args.append(kwargs[m_args[i]])
    elif arg_num < len(args):
        new_args.append(args[arg_num++])
```

Roughly something like this. There should be no else here - basically if function makes it to else - there are required arguments missing, so python will raise an error (but at this point it already should raise it if we called function with these params)

**Comment by 23doors** [comment]
Sorry for this long comment ;)

**Comment by thadeusb** [comment]
That last one was a doosey !

It gets tricky to delete memoized values by kwargs, it requires a reference to the function be passed in to delete_memoized so it can perform the kwarg->arg transition.

All unit tests are passing now, however I am afraid this could break backwards compatibility.

Commit: 6f7c6fbf94f3e208c8ae09d36fc7e2ad234fb941

**Comment by thadeusb** [comment]
Would you mind writing documentation on the use case for make_name?

**Comment by 23doors** [comment]
Nice touch with unless parameter!

I thought about it and I think that breaking backwards compatibility is inevitable.

For starters:
memoized stores f.**name** as a key (or make name) which means that:

```
module1:
@memoize()
def func1(id):
    return id

module2:
@memoize()
def func1(magic, def_arg='aa')
    return magic*2

val1 = module1.func1(1)
val2 = module2.func1(1)

val1 == val2
```

And this is default behavior, without make_name (which could prevent this).
Now, the only way to go around it is to:
memoize not with f.**name** but with '%s.%s' % (f.**module**, f.**name**) or at least add module to args/kwargs hash.

Now it may actually make more sense to allow using delete_memoized like this:
func1.delete_memoized(_args, *_kwargs)

OR similarly to celery project's task.delay and send_task functionality:
delete_memoized_by_name(function_name_and_module,  make_name, *args)

make_name is pretty much already required nonetheless. If it is customized, no way to deduce the cache_key.
Of course it shouldn't be a required param as some would simply prefer using:
delete_memoized_by_name(name_returned_by_my_make_name, *args)

We could simply require function_name_and_module to be valid, import the function and call delete_memoized for it.
I'm guessing that delete_memoized(function_name_and_module, _args, *_kwargs) - should do exactly that. But delete_memoized_by_name as an additional way of deleting keys wouldn't be that bad idea. It allows us to clear memoization of functions we don't want to import (for all kinds of reasons) or can't import.

About docs - I really hate writing documentation, hope you don't mind ;)
